src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -316,20 +316,12 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
    *   - `max(targetLen, arrayLen * 2, DefaultInitialSize)`.
    *   - Throws an exception if `targetLen` exceeds `VM_MaxArraySize` or is negative (overflow).
    */
-  private[mutable] def resizeUp(arrayLen: Int, targetLen: Int): Int = {
-    def checkArrayLengthLimit(): Unit =
-      if (targetLen > VM_MaxArraySize)
-        throw new Exception(s"Array of array-backed collection exceeds VM length limit of $VM_MaxArraySize. Requested length: $targetLen; current length: $arrayLen")
-      else if (targetLen < 0)
-        throw new Exception(s"Overflow while resizing array of array-backed collection. Requested length: $targetLen; current length: $arrayLen; increase: ${targetLen - arrayLen}")
-
-    if (targetLen > 0 && targetLen <= arrayLen) -1
-    else {
-      checkArrayLengthLimit()
-      if (arrayLen > VM_MaxArraySize / 2) VM_MaxArraySize
-      else math.max(targetLen, math.max(arrayLen * 2, DefaultInitialSize))
-    }
-  }
+  private[mutable] def resizeUp(arrayLen: Int, targetLen: Int): Int =
+    if (targetLen >= 0 && targetLen <= arrayLen) -1
+    else if (targetLen < 0) throw new Exception(s"Overflow while resizing array of array-backed collection. Requested length: $targetLen; current length: $arrayLen; increase: ${targetLen - arrayLen}")
```

Comment:
```suggestion
    if (targetLen < 0) throw new Exception(s"Overflow while resizing array of array-backed collection. Requested length: $targetLen; current length: $arrayLen; increase: ${targetLen - arrayLen}")
    else if (targetLen <= arrayLen) -1
```

I think it would be more readable to reorder the cases, so that `targetLen` overflow is tested only once. It might even have some (probably unmeasurable) performance gain - the test was already done and there is no need to do it twice.